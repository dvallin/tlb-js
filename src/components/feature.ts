import { Color } from '../renderer/color'
import { gray, primary } from '../renderer/palettes'
import { TlbWorld } from '../tlb'

import { FeatureComponent } from './feature'
import { PositionComponent } from './position'
import { WorldMap } from '../resources/world-map'
import { Vector } from '../spatial'
import { Entity } from '../ecs/entity'
import { strangeSymbols } from '../symbols'

export type FeatureType = keyof typeof features
export interface FeatureComponent {
  type: FeatureType
}

export interface Feature {
  character: string
  diffuse: Color

  blocking: boolean
  lightBlocking: boolean
  description: string
}

export const features = {
  wall: {
    character: '#',
    diffuse: gray[3],
    blocking: true,
    lightBlocking: true,
    description: 'a wall',
  },
  corridor: {
    character: '.',
    diffuse: gray[0],
    blocking: false,
    lightBlocking: false,
    description: 'a corridor',
  },
  room: {
    character: '.',
    diffuse: primary[1],
    blocking: false,
    lightBlocking: false,
    description: 'floor of a room',
  },
  locker: {
    character: strangeSymbols[16],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    description: 'a locker',
  },
  trash: {
    character: strangeSymbols[21],
    diffuse: gray[2],
    blocking: false,
    lightBlocking: false,
    description: 'some trash',
  },
  door: {
    character: strangeSymbols[27],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    description: 'a door',
  },
  hub: {
    character: '.',
    diffuse: primary[0],
    blocking: false,
    lightBlocking: false,
    description: 'floor of a hub',
  },
  player: {
    character: '@',
    diffuse: primary[0],
    blocking: true,
    lightBlocking: true,
    description: 'you',
  },
  guard: enemy('g', 'a guard'),
  eliteGuard: eliteEnemy('g', 'a very strong guard'),
}

function enemy(character: string, description: string): Feature {
  return {
    character,
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    description,
  }
}

function eliteEnemy(character: string, description: string): Feature {
  return {
    character,
    diffuse: primary[3],
    blocking: true,
    lightBlocking: true,
    description,
  }
}

export function getFeature(world: TlbWorld, entity: number): Feature | undefined {
  const feature = world.getComponent<FeatureComponent>(entity, 'feature')
  if (feature) {
    return features[feature.type]
  }
  return undefined
}

export function createFeature(world: TlbWorld, map: WorldMap, position: Vector, type: FeatureType): Entity {
  const e = map.getTile(position)
  if (e !== undefined) {
    throw new Error(`there is already a tile at ${position.key}`)
  }

  const entity = world
    .createEntity()
    .withComponent<PositionComponent>('position', { position })
    .withComponent<FeatureComponent>('feature', { type }).entity
  map.setTile(position, entity)
  return entity
}
