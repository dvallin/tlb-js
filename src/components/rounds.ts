export interface TakeTurnComponent {
  movements: number
  actions: number
}
