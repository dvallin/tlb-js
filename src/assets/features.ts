import { Feature } from '../components/feature'
import { gray, primary } from '../renderer/palettes'
import { strangeSymbols } from '../symbols'

function enemy(character: string, name: string): Feature {
  return {
    character,
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    name,
  }
}

function eliteEnemy(character: string, name: string): Feature {
  return {
    character,
    diffuse: primary[3],
    blocking: true,
    lightBlocking: true,
    name,
  }
}

const featuresDefinition = {
  wall: {
    character: '#',
    diffuse: gray[3],
    blocking: true,
    lightBlocking: true,
    name: 'wall',
  },
  corridor: {
    character: '.',
    diffuse: gray[0],
    blocking: false,
    lightBlocking: false,
    name: 'corridor',
  },
  room: {
    character: '.',
    diffuse: primary[1],
    blocking: false,
    lightBlocking: false,
    name: 'floor',
  },
  locker: {
    character: strangeSymbols[16],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    name: 'locker',
  },
  trash: {
    character: strangeSymbols[21],
    diffuse: gray[2],
    blocking: false,
    lightBlocking: false,
    name: 'trash',
  },
  door: {
    character: strangeSymbols[27],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: true,
    name: 'door',
  },
  hub: {
    character: '.',
    diffuse: primary[0],
    blocking: false,
    lightBlocking: false,
    name: 'floor',
  },
  player: {
    character: '@',
    diffuse: primary[0],
    blocking: true,
    lightBlocking: true,
    name: 'you',
  },
  loot: {
    character: 'l',
    diffuse: primary[1],
    blocking: false,
    lightBlocking: false,
    name: 'some loot',
  },
  table: {
    character: strangeSymbols[3],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: false,
    name: 'a table',
  },
  guard: enemy('g', 'guard'),
  eliteGuard: eliteEnemy('g', 'elite guard'),
}
export type FeatureType = keyof typeof featuresDefinition
export const features: { [key in FeatureType]: Feature } = featuresDefinition

const generatorsDefinition = {
  block: (index: number) => ({
    character: strangeSymbols[[7, 5, 17, 15][index]],
    diffuse: primary[1],
    blocking: true,
    lightBlocking: false,
    name: 'a block of concrete',
  }),
}
export type FeatureGeneratorsType = keyof typeof generatorsDefinition
export const generators: { [key in FeatureGeneratorsType]: (index: number) => Feature } = generatorsDefinition
