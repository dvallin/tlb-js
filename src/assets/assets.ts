import { Asset, Cover } from '../components/asset'
import { Vector } from '../spatial'
import { features, generators } from './features'
import { Feature } from '../components/feature'
import { DialogType } from './dialogs'

const s1x1 = new Vector([1, 1])
const s1x2 = new Vector([1, 2])
const s3x1 = new Vector([3, 1])
const s2x2 = new Vector([2, 2])
function asset(name: string, size: Vector, cover: Cover, dialog: DialogType | undefined, feature: (index: number) => Feature): Asset {
  return { name, size, cover, hasInventory: true, dialog, feature }
}

const assetsDefinition = {
  door: asset('metal door', s3x1, 'full', undefined, () => features['door']),
  locker: asset('a locker', s1x1, 'full', undefined, () => features['locker']),
  trash: asset('some trash', s1x1, 'none', undefined, () => features['trash']),
  loot: asset('a dead body', s1x1, 'none', undefined, () => features['loot']),
  table: asset('a table', s1x2, 'partial', undefined, () => features['table']),
  generator: asset('a generator', s2x2, 'full', undefined, generators['block']),
  elevator: asset('an elevator', s1x1, 'none', 'elevator', generators['elevator']),
}

export type AssetType = keyof typeof assetsDefinition
export const assets: { [key in AssetType]: Asset } = assetsDefinition
