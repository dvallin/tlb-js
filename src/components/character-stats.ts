import { ItemKind } from './items'

export type CharacterStatsType = keyof typeof characterStats
export interface CharacterStatsComponent {
  type: CharacterStatsType
  current: CharacterStats
}

export function createCharacterStatsComponent(type: CharacterStatsType): CharacterStatsComponent {
  return { type, current: JSON.parse(JSON.stringify(characterStats[type])) }
}

export type bodyPartType = 'head' | 'arm' | 'torso' | 'leg'

export interface BodyPart {
  type: bodyPartType
  itemAttachments: ItemKind[]
  health: number
}

export interface CharacterStats {
  bodyParts: { [key: string]: BodyPart }
  strength: number
  movement: number
  actions: number
  aim: number
}

function humanoidBodyParts(health: number): { [key: string]: BodyPart } {
  return {
    head: { type: 'head', itemAttachments: ['helmet'], health },
    leftArm: { type: 'arm', itemAttachments: ['glove', 'weapon'], health },
    rightArm: { type: 'arm', itemAttachments: ['glove', 'weapon'], health },
    torso: { type: 'torso', itemAttachments: ['armor'], health },
    leftLeg: { type: 'leg', itemAttachments: ['pants', 'boots'], health },
    rightLeg: { type: 'leg', itemAttachments: ['pants', 'boots'], health },
  }
}

export const characterStats = {
  player: {
    bodyParts: humanoidBodyParts(14),
    strength: 5,
    movement: 3,
    actions: 3,
    aim: 7,
  },
  guard: {
    bodyParts: humanoidBodyParts(8),
    strength: 5,
    movement: 3,
    actions: 3,
    aim: 3,
  },
  eliteGuard: {
    bodyParts: humanoidBodyParts(12),
    strength: 8,
    movement: 2,
    actions: 5,
    aim: 6,
  },
}
export const characterStatsTypeguard: { [key: string]: CharacterStats } = characterStats

export function speed(stats: CharacterStatsComponent): number {
  return stats.current.movement / 13
}
