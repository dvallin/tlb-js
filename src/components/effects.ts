import { Entity } from '../ecs/entity'
import { bodyPartType } from './character-stats'

export type effect = 'damage' | 'kill' | 'stun' | 'bleed' | 'confuse' | 'defend' | 'immobilize'

export interface Effect {
  type: effect
  global: boolean
  negated: boolean

  restricted?: bodyPartType[]
  duration?: number
  value?: number
}

export interface ActiveEffectsComponent {
  effects: {
    effect: Effect
    bodyPart?: string
  }[]
}

export interface EffectComponent {
  effect: Effect
  source: Entity
  target: Entity
  bodyParts?: string[]
}

export function damage(value: number, restricted: bodyPartType[] = ['torso']): Effect {
  return { type: 'damage', global: false, negated: false, value, restricted }
}

export function confuse(duration: number): Effect {
  return { type: 'confuse', global: true, negated: false, duration }
}

export function defend(value: number, duration: number): Effect {
  return { type: 'defend', global: true, negated: false, duration, value }
}

export function stun(duration: number): Effect {
  return { type: 'stun', global: true, negated: false, duration }
}

export function immobilize(duration: number): Effect {
  return { type: 'immobilize', global: true, negated: false, duration }
}

export function kill(): Effect {
  return { type: 'kill', global: true, negated: false }
}
