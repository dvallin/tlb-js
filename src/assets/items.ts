import { Effect } from '../components/effects'
import { ActionType } from './actions'
import { Item, WearableKind } from '../components/items'

function weapon(name: string, description: string, weight: number, effects: Effect[], actions: ActionType[], attachments: number): Item {
  return { kind: 'weapon', name, description, actions, effects, weight, attachments }
}

function consumable(name: string, description: string, weight: number, effects: Effect[]): Item {
  return { kind: 'consumable', attachments: 0, actions: ['consume'], name, description, effects, weight }
}

function equipment(
  kind: WearableKind,
  name: string,
  description: string,
  weight: number,
  effects: Effect[],
  actions: ActionType[],
  attachments: number
): Item {
  return { kind, name, description, actions, effects, weight, attachments }
}

export const itemsDefinition = {
  nailGun: weapon('nail gun', 'a tool to drive nails into wood, control panels or bodies.', 5, [], ['overcharge', 'hitAndRun', 'bolt'], 1),
  rifle: weapon('rifle', 'a regular rifle.', 5, [], ['shoot'], 2),
  sniperRifle: weapon('sniper rifle', '', 6, [], ['headshot', 'shoot'], 2),
  deathPill: consumable('death pill', 'a pill that kills you', 1, [{ type: 'kill', global: true, negated: false }]),
  bandages: consumable('bandages', 'a bandage to stop bleeding', 1, [{ type: 'bleed', global: false, negated: true }]),
  leatherJacket: equipment(
    'armor',
    'leather jacket',
    'a nice leather jacket',
    3,
    [{ type: 'defend', global: false, negated: false, value: 1 }],
    ['tighten'],
    1
  ),
  bootsOfStriding: equipment(
    'boots',
    'boots of striding',
    'Original Boots Of Striding TM, authentic D&D merch but not at all mint anymore',
    3,
    [],
    ['strideAndSlip'],
    2
  ),
}
export type ItemType = keyof typeof itemsDefinition
export const items: { [key in ItemType]: Item } = itemsDefinition
