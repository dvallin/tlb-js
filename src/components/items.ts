import { ActionType } from './action'
import { Entity } from '../ecs/entity'
import { Effect } from './effects'

export type WearableKind = 'glove' | 'helmet' | 'armor' | 'pants' | 'boots'
export type ItemKind = 'consumable' | 'weapon' | WearableKind

export interface Item {
  kind: ItemKind
  name: string
  description: string
  actions: ActionType[]
  effects: Effect[]
  weight: number
  attachments: number
}

export type ItemType = keyof typeof items
export const items = {
  nailGun: weapon('nail gun', 'a tool to drive nails into wood, control panels or bodies.', 5, [], ['overcharge', 'hitAndRun', 'bolt'], 1),
  rifle: weapon('rifle', 'a regular rifle.', 5, [], ['shoot'], 2),
  deathPill: consumable('death pill', 'a pill that kills you', 1, [{ type: 'kill', global: true, negated: false }]),
  bandages: consumable('bandages', 'a bandage to stop bleeding', 1, [{ type: 'bleed', global: false, negated: true }]),
  leatherJacket: equipment(
    'armor',
    'leather jacket',
    'a nice leather jacket',
    3,
    [{ type: 'defend', global: false, negated: false, value: 1 }],
    ['tighten'],
    1
  ),
  bootsOfStriding: equipment(
    'boots',
    'boots of striding',
    'Original Boots Of Striding TM, authentic D&D merch but not at all mint anymore',
    3,
    [],
    ['strideAndSlip'],
    2
  ),
}
export const itemsTypeguard: { [key: string]: Item } = items

export interface InventoryComponent {
  content: Entity[]
}

export interface ItemComponent {
  type: ItemType
}

export interface EquipmentAttachement {
  entity: Entity

  bodyParts: string[]
}

export interface EquipedItemsComponent {
  equipment: EquipmentAttachement[]
}

function weapon(name: string, description: string, weight: number, effects: Effect[], actions: ActionType[], attachments: number): Item {
  return { kind: 'weapon', name, description, actions, effects, weight, attachments }
}

function consumable(name: string, description: string, weight: number, effects: Effect[]): Item {
  return { kind: 'consumable', attachments: 0, actions: ['consume'], name, description, effects, weight }
}

function equipment(
  kind: WearableKind,
  name: string,
  description: string,
  weight: number,
  effects: Effect[],
  actions: ActionType[],
  attachments: number
): Item {
  return { kind, name, description, actions, effects, weight, attachments }
}
