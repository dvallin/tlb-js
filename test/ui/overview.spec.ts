import { Overview } from '../../src/ui/overview'
import { World } from '../../src/ecs/world'
import { TlbWorld } from '../../src/tlb'
import { mockInput, mockComponent, mockReturnValue, mockImplementation } from '../mocks'
import { Vector } from '../../src/spatial'
import { InventoryComponent, ItemComponent, EquipedItemsComponent } from '../../src/components/items'
import { Storage } from '../../src/ecs/storage'
import { Entity } from '../../src/ecs/entity'
import { Input } from '../../src/resources/input'
import { CharacterStatsComponent, createCharacterStatsComponent } from '../../src/components/character-stats'
import { ActiveEffectsComponent } from '../../src/components/effects'
import { TakeTurnComponent } from '../../src/components/rounds'
import { FeatureComponent } from '../../src/components/feature'
import { takeSnapshot, MockRenderer, createMockRenderer, clear } from './mock-display'
import { features } from '../../src/assets/features'

const mockItems: { [idx: number]: ItemComponent } = {
  2: { type: 'bandages' },
  3: { type: 'deathPill' },
}
const mockFeatures: { [idx: number]: FeatureComponent } = {
  4: { feature: () => features['eliteGuard'] },
  5: { feature: () => features['guard'] },
}

describe('ActionSelector', () => {
  let inventory: Storage<InventoryComponent>
  let item: Storage<ItemComponent>
  let equiped: Storage<EquipedItemsComponent>

  let stats: Storage<CharacterStatsComponent>
  let features: Storage<FeatureComponent>
  let activeEffects: Storage<ActiveEffectsComponent>

  let takeTurn: Storage<TakeTurnComponent>
  let tookTurn: Storage<{}>
  let waitTurn: Storage<{}>

  let overview: Overview
  let world: TlbWorld
  let renderer: MockRenderer
  let input: Input
  beforeEach(() => {
    overview = new Overview(0, 1, new Vector([0, 0]), 30)
    world = new World()
    input = mockInput(world)

    stats = mockComponent(world, 'character-stats')
    features = mockComponent(world, 'feature')
    activeEffects = mockComponent(world, 'active-effects')

    takeTurn = mockComponent(world, 'take-turn')
    tookTurn = mockComponent(world, 'took-turn')
    waitTurn = mockComponent(world, 'wait-turn')

    inventory = mockComponent(world, 'inventory')
    mockReturnValue<InventoryComponent>(inventory.get, { content: [] })
    renderer = createMockRenderer()
    item = mockComponent(world, 'item')
    equiped = mockComponent(world, 'equiped-items')
  })

  it('displays inventory', () => {
    mockReturnValue<InventoryComponent>(inventory.get, { content: [2, 3] })
    mockReturnValue<EquipedItemsComponent>(equiped.get, { equipment: [{ entity: 2, bodyParts: ['head', 'torso'] }] })
    mockImplementation(item.get, (e: Entity) => mockItems[e])

    overview.update(world)
    overview.render(renderer.renderer)

    takeSnapshot(renderer)
  })

  it('collapses inventory', () => {
    mockReturnValue<InventoryComponent>(inventory.get, { content: [2, 3] })
    mockReturnValue<EquipedItemsComponent>(equiped.get, { equipment: [{ entity: 2, bodyParts: ['head', 'torso'] }] })
    mockImplementation(item.get, (e: Entity) => mockItems[e])

    input.position = { x: 3, y: 0 }
    input.mousePressed = true
    overview.update(world)
    overview.render(renderer.renderer)

    clear(renderer)
    overview.update(world)
    overview.render(renderer.renderer)

    takeSnapshot(renderer)
  })

  it('displays body parts', () => {
    mockReturnValue<CharacterStatsComponent>(stats.get, createCharacterStatsComponent('player'))
    mockReturnValue<ActiveEffectsComponent>(activeEffects.get, {
      effects: [
        { effect: { type: 'bleed', global: false, negated: false }, bodyPart: 'torso' },
        { effect: { type: 'confuse', global: true, negated: false } },
      ],
    })

    overview.update(world)
    overview.render(renderer.renderer)

    takeSnapshot(renderer)
  })

  it('displays turn info', () => {
    mockReturnValue<CharacterStatsComponent>(stats.get, createCharacterStatsComponent('player'))
    mockReturnValue<ActiveEffectsComponent>(activeEffects.get, { effects: [] })
    mockReturnValue<TakeTurnComponent>(takeTurn.get, { movements: 1, actions: 2 })
    mockImplementation(features.get, (e: Entity) => mockFeatures[e])
    mockImplementation(waitTurn.foreach, (f: (entity: Entity) => void) => {
      f(4)
    })
    mockImplementation(tookTurn.foreach, (f: (entity: Entity) => void) => {
      f(5)
    })

    overview.update(world)
    overview.render(renderer.renderer)

    takeSnapshot(renderer)
  })
})
