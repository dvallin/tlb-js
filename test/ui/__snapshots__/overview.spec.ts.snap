// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionSelector collapses inventory 1`] = `
"------------------/inventory/-
"
`;

exports[`ActionSelector displays body parts 1`] = `
"-----------------/body parts/-
|head    |||||               |
|leftArm |||||               |
|rightArm|||||               |
|torso   ||||| b             |
|leftLeg |||||               |
|rightLeg|||||               |
|effects       c             |
------------------------------
------------------/inventory/-
------------------------------
"
`;

exports[`ActionSelector displays inventory 1`] = `
"------------------/inventory/-
|bandages @ head&torso       |
|death pill                  |
------------------------------
"
`;

exports[`ActionSelector displays turn info 1`] = `
"-----------------/body parts/-
|head    |||||               |
|leftArm |||||               |
|rightArm|||||               |
|torso   |||||               |
|leftLeg |||||               |
|rightLeg|||||               |
|effects                     |
------------------------------
------------------/inventory/-
------------------------------
-----------------------/turn/-
|2/3 AP 1/3 MP               |
|g guard                     |
|g elite guard               |
------------------------------
"
`;
